File: ros2_3d_C_ws/src/racer_3d_cpp/src/allocation.cpp
#include "racer_3d_cpp/allocation.hpp"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <utility>
#include <vector>

namespace racer_3d_cpp {
namespace {

double distance(const Point3 &first, const Point3 &second,
                const HGridDistanceFunction &distance_function) {
  if (distance_function) {
    const double value = distance_function(first, second);
    if (std::isfinite(value)) {
      return value;
    }
  }
  return (second - first).norm();
}

double routeCost(const std::vector<HGridCell3D> &route, const Point3 &start,
                 const HGridDistanceFunction &distance_function) {
  Point3 cursor = start;
  double result = 0.0;
  for (const auto &cell : route) {
    result += distance(cursor, cell.centroid, distance_function);
    cursor = cell.centroid;
  }
  return result;
}

std::vector<HGridCell3D> nearestRoute(
    const std::vector<HGridCell3D> &cells, const Point3 &start,
    const HGridDistanceFunction &distance_function) {
  std::vector<HGridCell3D> remaining = cells;
  std::vector<HGridCell3D> result;
  result.reserve(cells.size());
  Point3 cursor = start;
  while (!remaining.empty()) {
    std::size_t selected = 0;
    double best =
        distance(cursor, remaining.front().centroid, distance_function);
    for (std::size_t index = 1; index < remaining.size(); ++index) {
      const double candidate =
          distance(cursor, remaining[index].centroid, distance_function);
      if (candidate < best) {
        selected = index;
        best = candidate;
      }
    }
    result.push_back(remaining[selected]);
    cursor = remaining[selected].centroid;
    remaining.erase(remaining.begin() +
                    static_cast<std::ptrdiff_t>(selected));
  }
  return result;
}

struct ExactRoute {
  double cost{};
  std::vector<int> indices;
};

std::vector<ExactRoute> allExactOpenRoutes(
    const std::vector<HGridCell3D> &cells, const Point3 &start,
    const HGridDistanceFunction &distance_function) {
  const std::size_t count = cells.size();
  const std::size_t route_count = std::size_t{1} << count;
  std::vector<ExactRoute> routes(route_count);
  if (count == 0) {
    return routes;
  }

  std::vector<double> start_cost(count, 0.0);
  std::vector<std::vector<double>> pair_cost(
      count, std::vector<double>(count, 0.0));
  for (std::size_t first = 0; first < count; ++first) {
    start_cost[first] =
        distance(start, cells[first].centroid, distance_function);
    for (std::size_t second = first + 1; second < count; ++second) {
      const double value = distance(cells[first].centroid,
                                    cells[second].centroid,
                                    distance_function);
      pair_cost[first][second] = value;
      pair_cost[second][first] = value;
    }
  }

  const double infinity = std::numeric_limits<double>::infinity();
  std::vector<std::vector<double>> dynamic(
      route_count, std::vector<double>(count, infinity));
  std::vector<std::vector<int>> parent(route_count,
                                       std::vector<int>(count, -1));
  for (std::size_t mask = 1; mask < route_count; ++mask) {
    for (std::size_t last = 0; last < count; ++last) {
      const std::size_t bit = std::size_t{1} << last;
      if ((mask & bit) == 0) {
        continue;
      }
      const std::size_t previous_mask = mask ^ bit;
      if (previous_mask == 0) {
        dynamic[mask][last] = start_cost[last];
        continue;
      }
      double best = infinity;
      int best_previous = -1;
      for (std::size_t candidate = 0; candidate < count; ++candidate) {
        if ((previous_mask & (std::size_t{1} << candidate)) == 0) {
          continue;
        }
        const double value =
            dynamic[previous_mask][candidate] + pair_cost[candidate][last];
        // Python's min((cost, index), ...) chooses the lower index on ties.
        if (value < best ||
            (value == best &&
             static_cast<int>(candidate) < best_previous)) {
          best = value;
          best_previous = static_cast<int>(candidate);
        }
      }
      dynamic[mask][last] = best;
      parent[mask][last] = best_previous;
    }

    double best_cost = infinity;
    int final = -1;
    for (std::size_t last = 0; last < count; ++last) {
      if ((mask & (std::size_t{1} << last)) == 0) {
        continue;
      }
      const double value = dynamic[mask][last];
      if (value < best_cost ||
          (value == best_cost && static_cast<int>(last) < final)) {
        best_cost = value;
        final = static_cast<int>(last);
      }
    }

    std::vector<int> indices;
    std::size_t cursor_mask = mask;
    int cursor = final;
    while (cursor >= 0) {
      indices.push_back(cursor);
      const int previous =
          parent[cursor_mask][static_cast<std::size_t>(cursor)];
      cursor_mask ^= std::size_t{1} << static_cast<std::size_t>(cursor);
      cursor = previous;
    }
    std::reverse(indices.begin(), indices.end());
    routes[mask] = ExactRoute{best_cost, std::move(indices)};
  }
  return routes;
}

std::vector<std::string> cellIds(
    const std::vector<HGridCell3D> &cells) {
  std::vector<std::string> ids;
  ids.reserve(cells.size());
  for (const auto &cell : cells) {
    ids.push_back(cell.id());
  }
  return ids;
}

std::vector<HGridCell3D> cellsForRoute(
    const std::vector<HGridCell3D> &cells,
    const std::vector<int> &indices) {
  std::vector<HGridCell3D> result;
  result.reserve(indices.size());
  for (const int index : indices) {
    result.push_back(cells[static_cast<std::size_t>(index)]);
  }
  return result;
}

}  // namespace

Partition3D capacityPartition(
    const std::vector<HGridCell3D> &cells, const Point3 &first_start,
    const Point3 &second_start, double imbalance,
    const HGridDistanceFunction &distance_function) {
  if (cells.empty()) {
    return {};
  }

  int total_demand = 0;
  for (const auto &cell : cells) {
    total_demand += cell.demand();
  }

  if (cells.size() <= 10) {
    const std::size_t route_count = std::size_t{1} << cells.size();
    const std::size_t full_mask = route_count - 1;
    const auto first_routes =
        allExactOpenRoutes(cells, first_start, distance_function);
    const auto second_routes =
        allExactOpenRoutes(cells, second_start, distance_function);

    bool have_best = false;
    Partition3D best;
    for (std::size_t mask = 0; mask < route_count; ++mask) {
      std::vector<HGridCell3D> first;
      std::vector<HGridCell3D> second;
      int first_demand = 0;
      for (std::size_t index = 0; index < cells.size(); ++index) {
        if ((mask & (std::size_t{1} << index)) != 0) {
          first.push_back(cells[index]);
          first_demand += cells[index].demand();
        } else {
          second.push_back(cells[index]);
        }
      }
      const int second_demand = total_demand - first_demand;
      const double allowed =
          std::max(1.0, imbalance * static_cast<double>(total_demand));
      const double balance_error =
          std::abs(static_cast<double>(first_demand - second_demand));
      const auto &first_route = first_routes[mask];
      const auto &second_route = second_routes[full_mask ^ mask];
      const double objective =
          first_route.cost + second_route.cost +
          0.02 * std::max(0.0, balance_error - allowed);

      if (!have_best || objective < best.cost) {
        best.first = cellIds(first);
        best.second = cellIds(second);
        best.firstRoute =
            cellIds(cellsForRoute(cells, first_route.indices));
        best.secondRoute =
            cellIds(cellsForRoute(cells, second_route.indices));
        best.firstDemand = first_demand;
        best.secondDemand = second_demand;
        best.cost = objective;
        have_best = true;
      }
    }
    return best;
  }

  // Deterministic bounded fallback for unusually large pairwise active sets.
  std::vector<HGridCell3D> ordered = cells;
  std::stable_sort(ordered.begin(), ordered.end(),
                   [](const HGridCell3D &first,
                      const HGridCell3D &second) {
                     return first.demand() > second.demand();
                   });
  std::vector<HGridCell3D> first;
  std::vector<HGridCell3D> second;
  int demand[2] = {0, 0};
  for (const auto &cell : ordered) {
    const double first_score =
        (cell.centroid - first_start).norm() + 0.01 * demand[0];
    const double second_score =
        (cell.centroid - second_start).norm() + 0.01 * demand[1];
    const int owner = second_score < first_score ? 1 : 0;
    (owner == 0 ? first : second).push_back(cell);
    demand[owner] += cell.demand();
  }
  const auto first_route =
      nearestRoute(first, first_start, distance_function);
  const auto second_route =
      nearestRoute(second, second_start, distance_function);
  Partition3D result;
  result.first = cellIds(first);
  result.second = cellIds(second);
  result.firstRoute = cellIds(first_route);
  result.secondRoute = cellIds(second_route);
  result.firstDemand = demand[0];
  result.secondDemand = demand[1];
  result.cost = routeCost(first_route, first_start, distance_function) +
                routeCost(second_route, second_start, distance_function);
  return result;
}

}  // namespace racer_3d_cpp

File: ros2_3d_py_ws/src/racer_3d_py/racer_3d_py/allocation.py
"""Pairwise capacity-constrained 3-D region allocation."""

from dataclasses import dataclass
import math
from typing import Callable, Dict, List, Optional, Sequence, Tuple

import numpy as np

from .hgrid import HGridCell3D


DistanceFunction = Callable[[Sequence[float], Sequence[float]], float]


@dataclass
class Partition3D:
    first: List[str]
    second: List[str]
    first_route: List[str]
    second_route: List[str]
    first_demand: int
    second_demand: int
    cost: float


def _route_cost(
    route: Sequence[HGridCell3D],
    start: Sequence[float],
    distance_function: Optional[DistanceFunction] = None,
) -> float:
    cursor = np.asarray(start, dtype=float)
    result = 0.0
    for cell in route:
        target = np.asarray(cell.centroid)
        result += _distance(cursor, target, distance_function)
        cursor = target
    return result


def _distance(
    first: Sequence[float],
    second: Sequence[float],
    distance_function: Optional[DistanceFunction],
) -> float:
    if distance_function is not None:
        value = float(distance_function(first, second))
        if math.isfinite(value):
            return value
    return float(np.linalg.norm(np.asarray(second) - np.asarray(first)))


def _nearest_route(
    cells: Sequence[HGridCell3D],
    start: Sequence[float],
    distance_function: Optional[DistanceFunction] = None,
) -> List[HGridCell3D]:
    remaining = list(cells)
    result = []
    cursor = np.asarray(start, dtype=float)
    while remaining:
        item = min(
            remaining,
            key=lambda cell: _distance(
                cursor, cell.centroid, distance_function
            ),
        )
        result.append(item)
        remaining.remove(item)
        cursor = np.asarray(item.centroid)
    return result


def _all_exact_open_routes(
    cells: Sequence[HGridCell3D],
    start: Sequence[float],
    distance_function: Optional[DistanceFunction],
) -> Dict[int, Tuple[float, List[HGridCell3D]]]:
    """Held-Karp solutions for every open route subset.

    RACER's pairwise problem is small.  Computing all subset routes once for
    each vehicle gives exact open-path costs for every CVRP partition, rather
    than evaluating a nearest-neighbour approximation.
    """

    count = len(cells)
    routes: Dict[int, Tuple[float, List[HGridCell3D]]] = {0: (0.0, [])}
    if not count:
        return routes
    pair_cost = np.zeros((count, count), dtype=float)
    start_cost = np.zeros(count, dtype=float)
    for first in range(count):
        start_cost[first] = _distance(
            start, cells[first].centroid, distance_function
        )
        for second in range(first + 1, count):
            value = _distance(
                cells[first].centroid,
                cells[second].centroid,
                distance_function,
            )
            pair_cost[first, second] = value
            pair_cost[second, first] = value
    dynamic: Dict[Tuple[int, int], float] = {}
    parent: Dict[Tuple[int, int], int] = {}
    for mask in range(1, 1 << count):
        for last in range(count):
            bit = 1 << last
            if not mask & bit:
                continue
            previous_mask = mask ^ bit
            if previous_mask == 0:
                dynamic[mask, last] = float(start_cost[last])
                parent[mask, last] = -1
                continue
            previous = min(
                (
                    dynamic[previous_mask, candidate]
                    + pair_cost[candidate, last],
                    candidate,
                )
                for candidate in range(count)
                if previous_mask & (1 << candidate)
            )
            dynamic[mask, last] = float(previous[0])
            parent[mask, last] = int(previous[1])
        cost, final = min(
            (dynamic[mask, last], last)
            for last in range(count)
            if mask & (1 << last)
        )
        indices = []
        cursor_mask, cursor = mask, final
        while cursor >= 0:
            indices.append(cursor)
            previous = parent[cursor_mask, cursor]
            cursor_mask ^= 1 << cursor
            cursor = previous
        indices.reverse()
        routes[mask] = (float(cost), [cells[index] for index in indices])
    return routes


def capacity_partition(
    cells: Sequence[HGridCell3D],
    first_start: Sequence[float],
    second_start: Sequence[float],
    imbalance: float = 0.20,
    distance_function: Optional[DistanceFunction] = None,
) -> Partition3D:
    """Solve small pairwise CVRP partitions exactly, then route in 3-D."""

    values = list(cells)
    total_demand = sum(cell.demand for cell in values)
    if not values:
        return Partition3D([], [], [], [], 0, 0, 0.0)
    best = None
    # Pairwise HGrid interactions are deliberately bounded. Exact enumeration
    # reproduces RACER's capacity objective for up to 10 active cells.
    masks = range(1 << len(values)) if len(values) <= 10 else ()
    first_routes = (
        _all_exact_open_routes(values, first_start, distance_function)
        if len(values) <= 10
        else {}
    )
    second_routes = (
        _all_exact_open_routes(values, second_start, distance_function)
        if len(values) <= 10
        else {}
    )
    full_mask = (1 << len(values)) - 1
    for mask in masks:
        first = [
            cell for index, cell in enumerate(values) if mask & (1 << index)
        ]
        second = [cell for cell in values if cell not in first]
        first_demand = sum(cell.demand for cell in first)
        second_demand = total_demand - first_demand
        allowed = max(1.0, imbalance * total_demand)
        balance_error = abs(first_demand - second_demand)
        first_route_cost, first_route = first_routes[mask]
        second_route_cost, second_route = second_routes[full_mask ^ mask]
        route_cost = first_route_cost + second_route_cost
        objective = route_cost + 0.02 * max(0.0, balance_error - allowed)
        candidate = (
            objective,
            [cell.id for cell in first],
            [cell.id for cell in second],
            [cell.id for cell in first_route],
            [cell.id for cell in second_route],
            first_demand,
            second_demand,
        )
        if best is None or candidate[0] < best[0]:
            best = candidate
    if best is None:
        # Bounded deterministic fallback for an unusually large active set.
        first, second = [], []
        demand = [0, 0]
        for cell in sorted(values, key=lambda item: -item.demand):
            distances = (
                float(np.linalg.norm(np.asarray(cell.centroid) - first_start)),
                float(np.linalg.norm(np.asarray(cell.centroid) - second_start)),
            )
            owner = min(
                (0, 1),
                key=lambda index: distances[index] + 0.01 * demand[index],
            )
            (first if owner == 0 else second).append(cell)
            demand[owner] += cell.demand
        first_route = _nearest_route(
            first, first_start, distance_function
        )
        second_route = _nearest_route(
            second, second_start, distance_function
        )
        best = (
            _route_cost(first_route, first_start, distance_function)
            + _route_cost(second_route, second_start, distance_function),
            [cell.id for cell in first],
            [cell.id for cell in second],
            [cell.id for cell in first_route],
            [cell.id for cell in second_route],
            demand[0],
            demand[1],
        )
    return Partition3D(
        first=best[1],
        second=best[2],
        first_route=best[3],
        second_route=best[4],
        first_demand=best[5],
        second_demand=best[6],
        cost=best[0],
    )
